### PR TITLE
feat(ui/dfn-list): mark exported and unused definitions

### LINKED ISSUE #904: Lint for definitions that are defined but never used
It should flag things that are defined but never used as potential errors.

### LINKED ISSUE #2686: Exported definitions should be marked
In `ui/dfn-list` (Definitions), exported definitions should be marked differently than non-exported definitions. The UI for marking is for discussion.

### PR DESCRIPTION
**Pull request by janiceshiu**
Closes #2686
Incorporates some part of #904

Thanks for your help, @sidvishnoi!

### CODE REVIEW

File: tests/spec/core/ui-spec.js [hunk 1]
```diff
@@ -89,4 +89,40 @@ describe("Core - UI", () => {
...
+    const dfnListButton = doc.getElementById("respec-button-definitions");
+    dfnListButton.click();
+    await new Promise(resolve => setTimeout(resolve));
+
+    const dfns = doc.querySelectorAll("ul.respec-dfn-list li");
+    expect(dfns.length).toBe(4);
+
+    const [dfnBar, dfnBaz, dfnFoo, dfnFoobar] = dfns;
+
+    expect(dfnBar.innerText.includes("(exported) (unused)")).toBeTrue();
```
Comment (sidvishnoi) [inline tests/spec/core/ui-spec.js]: Nit:
```suggestion
    expect(dfnBar.textContent).toContain("(exported) (unused)");
```

File: tests/spec/core/ui-spec.js [hunk 2]
```diff
@@ -89,4 +89,40 @@ describe("Core - UI", () => {
...
+
+    expect(dfnBaz.innerText.includes("(unused)")).toBeTrue();
+    expect(dfnBaz.innerText.includes("(exported)")).toBeFalse();
```
Comment (sidvishnoi) [inline tests/spec/core/ui-spec.js]:
```suggestion
    expect(dfnBaz.textContent).not.toContain("(exported)");
```

File: tests/spec/core/ui-spec.js (hunk 1 above)
Comment (janiceshiu) [reply tests/spec/core/ui-spec.js]: @sidvishnoi ah - was looking for something like `toContain` and didn't find it in the docs. thanks!

could you explain why you prefer `textContent` over `innerText`?

File: src/ui/dfn-list.js [hunk 3]
```diff
@@ -49,9 +49,34 @@ function show() {
...
   ui.freshModal(l10n.list_of_definitions, ul, button);
 }
+
+/**
+ * If a definition is exported, label it accordingly
+ * @param {HTMLElement} dfn a definition
+ */
+function labelDfnIfExported(dfn) {
+  if (dfn.hasAttribute("data-export")) {
+    return hyperHTML`(exported)`;
```
Comment (sidvishnoi) [inline src/ui/dfn-list.js]: So we can style this.
```suggestion
    return hyperHTML`<span class="exported">(exported)</span>`;
```

File: tests/spec/core/ui-spec.js (hunk 1 above)
Comment (sidvishnoi) [reply tests/spec/core/ui-spec.js]: `textContent` vs `innerText`: https://stackoverflow.com/a/35213639
tl;dr: textContent is more performant, innerText capabilities are not needed here. We use textContent in most (all?) tests, so it's also consistent.

Comment (janiceshiu) [reply tests/spec/core/ui-spec.js]: fixed, thanks for explaining!

File: src/core/link-to-dfn.js [hunk 4]
```diff
@@ -193,9 +193,18 @@ function findLinkTarget(target, anchor, titleToDfns, possibleExternalLinks) {
...
+  markDfnAsUsed(dfn);
   return true;
 }
 
+/**
+ * Mark a definition as used once it has been used at least once within a spec
+ * @param {HTMLElement} dfn a definition
+ */
+function markDfnAsUsed(dfn) {
+  dfn.setAttribute("definition-used", "");
```
Comment (saschanaz) [inline src/core/link-to-dfn.js]: I would create a weakmap and export it rather than adding an attribute.

File: src/ui/dfn-list.js [hunk 5]
```diff
@@ -49,9 +49,34 @@ function show() {
...
+    return null;
+  }
+}
+
+/**
+ * If a definition is unused, label it accordingly
+ * @param {HTMLElement} dfn a definition
+ */
+function labelDfnIfUnused(dfn) {
+  if (dfn.hasAttribute("definition-used")) {
```
Comment (sidvishnoi) [inline src/ui/dfn-list.js]: I would prefer if we do this test here only, and not in `core/link-to-dfn`, although it's a but more efficient to that way.
Why? We use `isUnused` here only, and a user may not even open the dfn-list, so it makes sense to keep that logic localized here (makes it hard to reason about) and not do unnecessary work in link-to-dfn.

We can do something like:
``` js
const isUnused = !document.querySelector(`a[href^="#${dfn.id}"]`);
if (isUnused) {
  return null;
}
return hyperHTML`<span class="unused">(unused)</span>`;
```

Comment (janiceshiu) [reply src/ui/dfn-list.js]: @sidvishnoi do we expect all definitions in specs to have an `id`? 

* the sample spec in `basic.html` only has `<dfn>Foo</dfn>`.
* `<a>Foo</a>` results in `<a href="#dom-foo" class="internalDFN" data-link-type="idl"><code>Foo</code></a>` after respec has done its thing. 
* thus, using code above would mean that terms referenced in other parts of the spec would be incorrectly marked as `unused`

Comment (sidvishnoi) [reply src/ui/dfn-list.js]: > do we expect all definitions in specs to have an id?

Yes. If user doesn't specify an id, ReSpec auto-generates one.

> thus, using code above would mean that terms referenced in other parts of the spec would be incorrectly marked as unused

As UI is enabled after all plugins have processed, we are certain that ID generation and linking have been done already.

Comment (janiceshiu) [reply src/ui/dfn-list.js]: @sidvishnoi ah i see the issue. XD your code should've been `if (isUnused) { return hyperHTML(unused) etc}` that's why my test started failing as well. XD

**Comment by janiceshiu** [comment]
Note: I thought that the changes in https://github.com/w3c/respec/pull/2711/commits/0a0eb9154eb3b2453aa113f1f3c76663264f5741 were just a refactoring. However, I noticed that those changes weren't a refactor - they also changed functionality.

for example, in the sample spec in `basic.html`,
* using the old method of marking the term with an attribute if `findLinkTarget` returns `true`
  * the terms `bar`, `Foo.constructor()` and `doTheFoo()` are marked as `(unused)`
* using the new method of `document.querySelector` for `a href`
  * no terms are marked as `(unused)`

@sidvishnoi mentioned he'd look into it. commenting here in case anyone else has an explanation. ^^

File: assets/dfn-panel.css [hunk 6]
```diff
@@ -3,6 +3,25 @@ dfn {
   cursor: pointer;
 }
 
+span.exported, span.unused {
```
Comment (sidvishnoi) [inline assets/dfn-panel.css]: These show go in https://github.com/w3c/respec/blob/develop/assets/ui.css

File: assets/ui.css [hunk 7]
```diff
@@ -426,3 +426,22 @@
...
+
+.respec-dfn-list li span.exported, span.unused {
+  margin-left: 8px;
+  padding: 0.1em;
+  text-align: center;
+  white-space: nowrap;
+}
+
+.respec-dfn-list li span.exported {
+  background: hsl(205.7, 70%, 96.1%);
```
Comment (sidvishnoi) [inline assets/ui.css]: Nit: Lets round-off all the colors.
```suggestion
  background: hsl(206, 70%, 96%);
```

File: src/ui/dfn-list.js [hunk 8]
```diff
@@ -49,9 +49,33 @@ function show() {
...
   ui.freshModal(l10n.list_of_definitions, ul, button);
 }
+
+/**
+ * If a definition is exported, label it accordingly
+ * @param {HTMLElement} dfn a definition
+ */
+function labelDfnIfExported(dfn) {
+  if (dfn.hasAttribute("data-export")) {
+    return hyperHTML`<span class="exported">(exported)</span>`;
```
Comment (sidvishnoi) [inline src/ui/dfn-list.js]: Maybe remove the parenthesis? 
```suggestion
    return hyperHTML`<span class="exported">exported</span>`;
```

File: assets/ui.css [hunk 9]
```diff
@@ -426,3 +426,22 @@
 #xref-ui:not(.ready) {
   background: url("https://respec.org/xref/loader.gif") no-repeat center;
 }
+
+.respec-dfn-list li span.exported, span.unused {
+  margin-left: 8px;
```
Comment (sidvishnoi) [inline assets/ui.css]: Nit: for consistency with rest of styles.
```suggestion
  margin-left: 0.5em;
```

File: assets/ui.css (hunk 7 above)
Comment (saschanaz) [reply assets/ui.css]:
```suggestion
  background: #eef6fc;
```

File: src/ui/dfn-list.js (hunk 8 above)
Comment (janiceshiu) [reply src/ui/dfn-list.js]: imho keeping the parenthesis is good in case the styling doesn't jive with whatever the user is using to read the specs. otherwise they might think that `exported` or `unused` is part of the definition name.

Comment (sidvishnoi) [reply src/ui/dfn-list.js]: I don't have a strong opinion on this. Happy to keep them, happy to remove them. 
BTW, if styling didn't work, most likely JS didn't work either.

Comment (janiceshiu) [reply src/ui/dfn-list.js]: hm. also thinking of colour-blindness and stuff. for example, there's a type of colour-blindness called `Achromatopsia` where vision is essentially monochromatic. folk like that would probably appreciate the parentheses.

File: assets/ui.css (hunk 7 above)
Comment (janiceshiu) [reply assets/ui.css]: changed the lot to hex in https://github.com/w3c/respec/pull/2711/commits/2e75a7b1b4d11d14de299f49fa1691915ddf1547

File: assets/ui.css [hunk 10]
```diff
@@ -427,6 +427,25 @@
   background: url("https://respec.org/xref/loader.gif") no-repeat center;
 }
 
+.respec-dfn-list li span.exported, span.unused {
```
Comment (sidvishnoi) [inline assets/ui.css]:
```suggestion
.respec-dfn-list li span.exported, .respec-dfn-list li span.unused {
```
or:
``` suggestion
.respec-dfn-list li span {
```

File: src/ui/dfn-list.js [hunk 11]
```diff
@@ -49,9 +49,33 @@ function show() {
...
+  }
+  return null;
+}
+
+/**
+ * If a definition is unused, label it accordingly
+ * @param {HTMLElement} dfn a definition
+ */
+function labelDfnIfUnused(dfn) {
+  const isUsed = document.querySelector(`a[href^="#${dfn.id}"]`);
```
Comment (sidvishnoi) [inline src/ui/dfn-list.js]: Nit: maybe use this for consistency with labelDfnIfExported
```js
  const isUsed = document.querySelector(`a[href^="#${dfn.id}"]`);
  if (!isUsed) {
    return hyperHTML`<span class="unused">(unused)</span>`;
  }
  return null;
```

Comment (janiceshiu) [reply src/ui/dfn-list.js]: hm. one `!` so i'm still okay with it. XD

if we're going with that, then for overall consistency, we should change `labelDfnIfExported/1` to
```javascript
function labelDfnIfExported(dfn) {
  const isExported =  dfn.hasAttribute("data-export");
  if (isExported) {
    return hyperHTML`<span class="exported">(exported)</span>`;
  }
  return null;
}
```

File: src/ui/dfn-list.js (hunk 8 above)
Comment (saschanaz) [reply src/ui/dfn-list.js]: > hm. also thinking of colour-blindness and stuff. for example, there's a type of colour-blindness called `Achromatopsia` where vision is essentially monochromatic. folk like that would probably appreciate the parentheses.

Does the same apply for GitHub labels? Does GitHub do something for them?

File: assets/ui.css [hunk 12]
```diff
@@ -427,6 +427,27 @@
...
+  margin-left: 0.5em;
+  padding: 0.1em;
+  text-align: center;
+  white-space: nowrap;
+  font-size: 90%;
+  border-radius: 0.2em;
+}
+
+.respec-dfn-list li span.exported {
+  background: #fcfbee;
```
Comment (saschanaz) [inline assets/ui.css]: image

Hard to tell the difference between unused in no-green/no-red accessibility emulation mode, we might want to adjust colors for better accessibility.

Comment (sidvishnoi) [reply assets/ui.css]: I found `#d1edfd` to be a good fit for background for `span.exported`. WDYT?

Comment (sidvishnoi) [reply assets/ui.css]: Wow. GitHub shows color preview in comments!

Comment (saschanaz) [reply assets/ui.css]: Yes, seems the following works:

```css
.respec-dfn-list li span.unused {
	background: #d1edfd;
	color: #040b1c;
	box-shadow: 0 0 0 0.125em #1ca5f940;
}
```

Comment (saschanaz) [reply assets/ui.css]: Can the selector be simplified? I think `.unused` is enough, or maybe `.unused-dfn` for less ambiguity.

Comment (sidvishnoi) [reply assets/ui.css]: I think `.respec-dfn-list` based scoping is good idea (makes it easy to keep related code together). We can remove the `li span` part.

**Comment by janiceshiu** [comment]
all updated! @saschanaz if all is good can you approve? thanks!

File: src/ui/dfn-list.js (hunk 8 above)
Comment (janiceshiu) [reply src/ui/dfn-list.js]: @saschanaz interesting question. i don't know!

Comment (saschanaz) [reply src/ui/dfn-list.js]: Can we go without parentheses for now? Because:

* Wikipedia says total color blindness is not just a normal colorblind but a more serious medical syndrome.
* Even if someone is total colorblind they should still be able to see gray background

So for now I don't think additional visual indication is needed.

File: assets/ui.css [hunk 13]
```diff
@@ -427,6 +427,27 @@
   background: url("https://respec.org/xref/loader.gif") no-repeat center;
 }
 
+.respec-dfn-list li span {
```
Comment (saschanaz) [inline assets/ui.css]: I would prefer using a proper classname here rather than `li span`, for example `.dfn-status`.

Comment (janiceshiu) [reply assets/ui.css]: @saschanaz makes sense. in case we do have other elements that also match `li span`. added `.dfn-status` in https://github.com/w3c/respec/pull/2711/commits/d5108429caa30bd82e42cc1852b33a35b036fa53!

Comment (janiceshiu) [reply assets/ui.css]: @saschanaz makes sense. in case we do have other elements that also match li span. added .dfn-status in d510842!

**Comment by sidvishnoi** [comment]
ReSpec v25.1.0 coming up!

**Comment by marcoscaceres** [comment]
Very cool, @janiceshiu! Nice one.
